<a href="https://colab.research.google.com/github/ralphcajipe/Deep-Learning-with-Python-2nd-edition/blob/main/2_MNIST_hello_world.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Modified National Institute of Standards and Technology dataset

### Problem: To classify grayscale images of handwritten digits (28 x 28 pixels) into their 10 categories (0 through 9).

**Loading the MNIST dataset in Keras**

In [2]:
from tensorflow.keras.datasets import mnist
(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

11501568/11490434 [==============================] - 0s 0us/step


**Training data:**

In [3]:
train_images.shape

(60000, 28, 28)

In [4]:
len(train_labels)

60000

In [5]:
train_labels

array([5, 0, 4, ..., 5, 6, 8], dtype=uint8)

**Test data:**

In [6]:
test_images.shape

(10000, 28, 28)

In [7]:
len(test_labels)

10000

In [8]:
test_labels

array([7, 2, 1, ..., 4, 5, 6], dtype=uint8)

**The network architecture**

In [9]:
from tensorflow import keras
from tensorflow.keras import layers
model = keras.Sequential([
    layers.Dense(512, activation="relu"),
    layers.Dense(10, activation="softmax")
])

**The compilation step**

In [11]:
model.compile(optimizer="rmsprop",
              loss="sparse_categorical_crossentropy",
              metrics=["accuracy"])

**Preparing the image data**

In [13]:
train_images = train_images.reshape((60000, 28 * 28))
train_images = train_images.astype("float32") / 255
test_images = test_images.reshape((10000, 28 * 28))
test_images = test_images.astype("float32") / 255

**"Fitting" the deep learning model**

In [14]:
model.fit(train_images, train_labels, epochs=5, batch_size=128)

Epoch 1/5
469/469 [==============================] - 6s 11ms/step - loss: 0.2562 - accuracy: 0.9262
Epoch 2/5
469/469 [==============================] - 5s 11ms/step - loss: 0.1037 - accuracy: 0.9688
Epoch 3/5
469/469 [==============================] - 5s 11ms/step - loss: 0.0684 - accuracy: 0.9791
Epoch 4/5
469/469 [==============================] - 5s 11ms/step - loss: 0.0501 - accuracy: 0.9847
Epoch 5/5
469/469 [==============================] - 5s 11ms/step - loss: 0.0370 - accuracy: 0.9887


Training-set accuracy: 98.9%

**Using the deep learning model to make predictions**

In [15]:
test_digits = test_images[0:10]
predictions = model.predict(test_digits)
predictions[0]

array([2.5092247e-08, 3.9475909e-10, 1.3601459e-06, 1.8806270e-05,
       8.2402600e-12, 2.9451694e-08, 1.7941699e-14, 9.9997592e-01,
       1.9641561e-08, 3.8495537e-06], dtype=float32)

Eeach number of index `i` in that array corresponds to the probability that digit image `test_digits[0]` belongs to class `i`.

This first test digit has the highest probability score (0.9999759, almost 1) at index 7, so according to our model, it must be a number 7:

In [21]:
predictions[0].argmax()

7

In [33]:
predictions[0][7]

0.9999759

In [35]:
print(f"{predictions[0][7]:.1%}")

100.0%


**Check that the test label agrees:**

In [22]:
test_labels[0]

7

It's really a number 7!

**Evaluating the model on new data**

In [36]:
test_loss, test_acc = model.evaluate(test_images, test_labels)
print(f"test_acc: {test_acc}")
print(f"{test_acc:.1%}")

313/313 [==============================] - 1s 3ms/step - loss: 0.0685 - accuracy: 0.9804
test_acc: 0.980400025844574
98.0%


Test-set accuracy: 98.0%